plugins/node/opentelemetry-instrumentation-aws-sdk/src/services/lambda.ts

```diff
@@ -0,0 +1,133 @@
+ * limitations under the License.
+ */
+import {
+  Span,
+  SpanKind,
+  Tracer,
+  diag,
+  SpanAttributes,
+} from '@opentelemetry/api';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import {
+  AwsSdkInstrumentationConfig,
+  NormalizedRequest,
+  NormalizedResponse,
+} from '../types';
+import { RequestMetadata, ServiceExtension } from './ServiceExtension';
+import { TextMapSetter, context, propagation } from '@opentelemetry/api';
+import { InvocationRequest } from '@aws-sdk/client-lambda';
+
+export class LambdaServiceExtension implements ServiceExtension {
+  requestPreSpanHook(request: NormalizedRequest): RequestMetadata {
+    const functionName = this.extractFunctionName(request.commandInput);
+
+    let spanAttributes: SpanAttributes = {};
+    let spanName: string | undefined;
+
+    switch (request.commandName) {
+      case 'Invoke':
+        spanAttributes = {
+          [SemanticAttributes.FAAS_INVOKED_NAME]: functionName,
+          [SemanticAttributes.FAAS_INVOKED_PROVIDER]: 'aws',
+        };
+        if (request.region) {
+          spanAttributes[SemanticAttributes.FAAS_INVOKED_REGION] =
+            request.region;
+        }
+        spanName = `${functionName} invoke`;
+        break;
+    }
+    return {
+      isIncoming: false,
+      spanAttributes,
+      spanKind: SpanKind.CLIENT,
+      spanName,
+    };
+  }
+
+  requestPostSpanHook = (request: NormalizedRequest) => {
+    switch (request.commandName) {
+      case 'Invoke':
+        {
+          request.commandInput = injectPropagationContext(
+            request.commandInput as InvocationRequest
+          );
+        }
+        break;
+    }
+  };
+
+  responseHook(
+    response: NormalizedResponse,
+    span: Span,
+    tracer: Tracer,
+    config: AwsSdkInstrumentationConfig
+  ) {
+    const operation = response.request.commandName;
+
+    if (operation === 'Invoke') {
+      span.setAttribute(SemanticAttributes.FAAS_EXECUTION, response.requestId);
+    }
+  }
+
+  extractFunctionName = (commandInput: Record<string, any>): string => {
+    return commandInput?.FunctionName;
+  };
+}
+class ContextSetter implements TextMapSetter<Record<string, any>> {
+  set(carrier: Record<string, any>, key: string, value: string): void {
+    try {
+      const parsedClientContext = carrier.ClientContext
+        ? JSON.parse(
+            Buffer.from(carrier.ClientContext, 'base64').toString('utf8')
+          )
+        : {};
+      const updatedPayload = {
+        ...parsedClientContext,
+        Custom: {
+          ...parsedClientContext.Custom,
+          [key]: value,
+        },
+      };
+      const encodedPayload = Buffer.from(
+        JSON.stringify(updatedPayload)
+      ).toString('base64');
+
+      // The length of client context is capped at 3583 bytes of base64 encoded data
+      // (https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_RequestSyntax)
+      if (encodedPayload.length > 3583) {
```

Comment:
If the function is invoked multiple times (for example, `traceparent` and `tracestate` for `W3CTraceContextPropagator`, or the 4 fields of `B3MultiPropagator` propagator), it is possible that only part of the context will be injected and some will be missing after the limit is reached.
For SQS / SNS, [the inject function is first checking if all context fields can be injected successfully and only then attempts any modifications to the request](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/986c3499b324ebdf49aabf35bc3711c91bb91ec8/plugins/node/opentelemetry-instrumentation-aws-sdk/src/services/MessageAttributes.ts#L74).

What do you think about using a similar approach here as well?

This will also be more performant, as currently, we decode the base64 string, JSON.parse it, then JSON stringify it and encode it back to base64, again and again for each field.

Reply:
I noted that approach, but wasn't sure if there was a precise enough equivalent to the 'will this send it over the limit' question, when we first don't know the length of the key + value (the SQS/SNS equivalent is for max number of keys, the values are free in length, so don't need checking) and we would have to use a bodgy encoded => base64 length mapping formula. I'll give it a try to see if I can get all tests passing but it might need a final check before setting just in case despite best endeavours it still has managed to go over the limit.

Reply:
What I imagined is something like this (haven't really run it, it's just a scatch):
```js
const injectPropagationContext = (
  invocationRequest: InvocationRequest
): InvocationRequest => {
  try {
    const propagationKeysValues = {};
    propagation.inject(context.active(), propagationKeysValues);

    const parsedClientContext = invocationRequest.ClientContext
    ? JSON.parse(
        Buffer.from(invocationRequest.ClientContext, 'base64').toString('utf8')
      )
    : {};

    const updatedClientContext = {
      ...parsedClientContext,
      Custom: {
        ...parsedClientContext.Custom,
        ...propagationKeysValues,
      },
    };

    const encodedClientContext = Buffer.from(
      JSON.stringify(updatedClientContext)
    ).toString('base64');

    // The length of client context is capped at 3583 bytes of base64 encoded data
    // (https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_RequestSyntax)
    if (encodedClientContext.length > 3583) {
      diag.warn(
        'lambda instrumentation: cannot set context propagation on lambda invoke parameters due to ClientContext length limitations.'
      );
      return invocationRequest;      
    }

    return { 
      ...invocationRequest,
      ClientContext: encodedClientContext,
    }

  } catch(err) {
    diag.debug(
      'lambda instrumentation: failed to set context propagation on ClientContext',
      err
    );
    return invocationRequest;
  }
};
```

This code will either inject the entire propagation data, or none of it. It will also `JSON.parse` + `Buffer.from` and `JSON.stringify` + `toString('base64')` the client context at most once, regardless of the number of propagation keys.
It will also create a copy of the request params, instead of modifying the user's object which might introduce side effects